### PR TITLE
Add "optimize image" blog setting

### LINKED ISSUE #5228: Improve the picture resize / optimize image option
We all agreed that the current resizing options are kind of lame. 
A single `Optimize Pictures` toggle seems the way to go.  Once enabled the picture will be resized to a pre-defined size, and compressed .85 or .90 quality.  Maybe offering advanced settings for power users.

screen-shot-2017-02-17-at-09-43-05

Note: Would be interesting to make the option apply only if bandwidth is limited (not on wifi).

The following tickets are related, but no need to work on them if we want to pursue this way.

Ref 1: https://github.com/wordpress-mobile/WordPress-Android/issues/4241
Ref 2: https://github.com/wordpress-mobile/WordPress-Android/issues/2819

### PR DESCRIPTION
**Pull request by daniloercoli**
Working on #5228 by adding a checkbox that will be used to optimize pictures on 3G connections only.

@oguzkocer Since site settings is your baby, would you mind to take a look and check if i'm doing it right?

- `Optimize pictures` & `Location settings` should be stored "per-site"basis. 
- `Optimize pictures` will be used to compress the picture a bit when enabled and on 3G connections.

***Edited***
This PR fixes #5228 by adding a simple `Optimize images` checkbox to site settings. Once enabled, and on 3G connections, the app resize the pictures to 3000px width (if the picture is wider) and always compress it with quality=85.
- GIF pictures are excluded from the optimization.
- We may want to add a `Custom/Advanced` options section, as shown in the original ticket, in the future. Depends on the users feedback on this feature.

### CODE REVIEW

**Comment by oguzkocer** [comment]
@daniloercoli I think it's @tonyr59h's baby, but I'd be happy to review it! I'll most likely do it tomorrow though, if it's not urgent :)

File: WordPress/src/main/java/org/wordpress/android/models/SiteSettingsModel.java [hunk 1]
```diff
@@ -133,14 +136,20 @@ public boolean equals(Object other) {
         SiteSettingsModel otherModel = (SiteSettingsModel) other;
 
         return localTableId == otherModel.localTableId &&
-                address.equals(otherModel.address) &&
-                username.equals(otherModel.username) &&
-                password.equals(otherModel.password) &&
-                title.equals(otherModel.title) &&
-                tagline.equals(otherModel.tagline) &&
+                ((address == null && otherModel.address == null) ||
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/models/SiteSettingsModel.java]: These conditionals feels better than what it used to be, but not sure if it's necessary. Is there an issue that this is fixing? If so, should we separate it into it's own PR?

Also, assuming we keep these changes, is there a reason other objects don't use the same pattern like `defaultPostFormat` for example?

Again, assuming the changes are kept, do you think we should introduce a helper method to simplify this a bit? That way it'll be easier to read and understand and also apply it to other objects. What do you think?

File: WordPress/src/main/java/org/wordpress/android/models/SiteSettingsModel.java [hunk 2]
```diff
@@ -158,8 +167,9 @@ public boolean equals(Object other) {
                 commentsRequireUserAccount == otherModel.commentsRequireUserAccount &&
                 commentAutoApprovalKnownUsers == otherModel.commentAutoApprovalKnownUsers &&
                 maxLinks == otherModel.maxLinks &&
-                holdForModeration != null && holdForModeration.equals(otherModel.holdForModeration) &&
-                blacklist != null && blacklist.equals(otherModel.blacklist);
+                holdForModeration != null && otherModel.holdForModeration != null
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/models/SiteSettingsModel.java]: I don't think `otherModel.holdForModeration != null` is necessary here. The other 2 cases will cover this case. If the `holdForModeration ` is not `null` and equal to `otherModel.holdForModeration`, then it means `otherModel.holdForModeration` is not `null` either.

We could keep it for verbosity though, so the devs reading the code doesn't need to think about it. It's up to you!

File: WordPress/src/main/java/org/wordpress/android/models/SiteSettingsModel.java [hunk 3]
```diff
@@ -158,8 +167,9 @@ public boolean equals(Object other) {
...
+                    && holdForModeration.equals(otherModel.holdForModeration) &&
+                blacklist != null && otherModel.blacklist != null && blacklist.equals(otherModel.blacklist);
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/models/SiteSettingsModel.java]: Same comment as above about the `holdForModeration`.

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsInterface.java [hunk 4]
```diff
@@ -282,6 +277,10 @@ public boolean getLocation() {
         return mSettings.location;
     }
 
+    public boolean getOptmizeImage() {
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsInterface.java]: Small spelling mistake here, should be `getOptimizeImage`.

Better yet, we could use `optimizedImage` as the property and `getOptimizedImage` for the getter, but that's not a big deal.

File: WordPress/src/main/java/org/wordpress/android/models/SiteSettingsModel.java (hunk 1 above)
Comment (daniloercoli) [inline WordPress/src/main/java/org/wordpress/android/models/SiteSettingsModel.java]: Right, will introduce a helper method.

**Comment by daniloercoli** [comment]
@oguzkocer This is ready for another review.  
There are same issues on uploading optimized pictures I still need to figured them out, but the migration code is there, as well as all the major things.

**Comment by daniloercoli** [comment]
Hey @mzorz, if you want to take a look at this PR it's almost ready for review.

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java [hunk 5]
```diff
@@ -139,9 +139,15 @@ public void onResponse(JSONObject response) {
                             // postFormats setting is not returned by this api call so copy it over
                             final Map<String, String> currentPostFormats = mSettings.postFormats;
 
+                            // Local settings
+                            boolean location = mSettings.location;
+                            boolean optimizeImaged = mSettings.optimizedImage;
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java]: There is a small typo in `optimizeImaged`.

**Comment by oguzkocer** [comment]
@daniloercoli I am getting the following error when I try to install this branch on top of `feature/fluxc-integration`:

```
Installation failed since the device already has a newer version of this application.
In order to proceed, you have to uninstall the existing application.

WARNING: Uninstalling will remove the application data!

Do you want to uninstall the existing application?
```
We released another alpha yesterday, so I am guessing all you need to do is merge `feature/fluxc-integration` into this branch. I am bringing it up because I was unable to test the migration and had to do  a fresh install instead.

**Comment by daniloercoli** [comment]
Thanks @oguzkocer. Merged `fluxc-integration` in it, and applied final changes.
@mzorz Your turn of review.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 6]
```diff
@@ -1537,6 +1537,25 @@ private boolean addMediaVisualEditor(Uri uri, boolean isVideo) {
...
         }
 
+        if (!isVideo && SiteSettingsInterface.getInterface(this, mSite, null).init(false).getOptimizedImage() &&
+               !NetworkUtils.isWiFiConnected(this)) {
+            // Not on WiFi and optimize image is set to ON
+            String optimizedPath = ImageUtils.optimizeImage(this, path);
+
+            if (optimizedPath == null) {
+                AppLog.e(T.EDITOR, "Optimized picture was null!");
+                //TODO: track analytics here
```
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: minor, very minor comment: please use space after the `//`. Like, `// TODO` instead of `//TODO`. I had maxime ask me for this once 😄

Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: Also, are you planning to add analytics tracking in this PR?

File: libs/utils/WordPressUtils/src/main/java/org/wordpress/android/util/ImageUtils.java [hunk 7]
```diff
@@ -495,7 +495,7 @@ private static boolean resizeImageAndWriteToStream(Context context,
         }
 
         Bitmap.CompressFormat fmt;
-        if (fileExtension != null && fileExtension.equalsIgnoreCase("png")) {
+        if (fileExtension != null && fileExtension.contains("png")) {
```
Comment (mzorz) [inline libs/utils/WordPressUtils/src/main/java/org/wordpress/android/util/ImageUtils.java]: why this change? can a PNG file be any other thing than `png`? Or the other way around, couldn't this change bring false positives in?

File: libs/utils/WordPressUtils/src/main/java/org/wordpress/android/util/ImageUtils.java [hunk 8]
```diff
@@ -523,17 +523,20 @@ private static boolean resizeImageAndWriteToStream(Context context,
 
         boolean result = bmpRotated.compress(fmt, quality, outStream);
         bmpRotated.recycle();
+        bmpResized.recycle();
```
Comment (mzorz) [inline libs/utils/WordPressUtils/src/main/java/org/wordpress/android/util/ImageUtils.java]: I think we don't even need to call recycle() as we have a minSdk of 16.
Check this https://developer.android.com/topic/performance/graphics/manage-memory.html

`Android 2.3 adds concurrent garbage collection, which means that the memory is reclaimed soon after a bitmap is no longer referenced.`

File: libs/utils/WordPressUtils/src/main/java/org/wordpress/android/util/ImageUtils.java [hunk 9]
```diff
@@ -545,47 +548,52 @@ public static String createResizedImageWithMaxWidth(Context context, String path
...
+        int selectedWidth = getImageSize(srcImageUri, context)[0];
+        if (selectedWidth == 0) {
+            // Can't read the src dimensions.
             return path;
         }
-
-        Uri imageUri = Uri.parse(path);
-        if (context == null || imageUri == null) {
-            return path;
+        final int maxImageWidth = 3000;
```
Comment (mzorz) [inline libs/utils/WordPressUtils/src/main/java/org/wordpress/android/util/ImageUtils.java]: If there's any reason why we chose the number 3000, can we make it explicit in a comment please?

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java (hunk 6 above)
Comment (daniloercoli) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: Not in this PR. Will add analytics later in its own PR. Just want to think a little bit more on it.

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java (hunk 5 above)
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java]: Agreed, the typo is still there

File: libs/utils/WordPressUtils/src/main/java/org/wordpress/android/util/ImageUtils.java (hunk 7 above)
Comment (daniloercoli) [inline libs/utils/WordPressUtils/src/main/java/org/wordpress/android/util/ImageUtils.java]: Ok, let me redo this, since it's not clear.

File: libs/utils/WordPressUtils/src/main/java/org/wordpress/android/util/ImageUtils.java (hunk 8 above)
Comment (daniloercoli) [inline libs/utils/WordPressUtils/src/main/java/org/wordpress/android/util/ImageUtils.java]: I just moved this line of code at the bottom of the method since was seeing some issues on older devices.
We can remove those calls, but don't think it's hurt having them there.

Comment (mzorz) [inline libs/utils/WordPressUtils/src/main/java/org/wordpress/android/util/ImageUtils.java]: I think we can safely leave it up to the Android system. May I suggest we bring someone else's opinion on this particular?

Comment (daniloercoli) [inline libs/utils/WordPressUtils/src/main/java/org/wordpress/android/util/ImageUtils.java]: Ok, removed them. We should check that the memory is actually freed after the method ends.

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java [hunk 10]
```diff
@@ -141,13 +141,13 @@ public void onResponse(JSONObject response) {
 
                             // Local settings
                             boolean location = mSettings.location;
-                            boolean optimizeImaged = mSettings.optimizedImage;
+                            boolean optimizedImaged = mSettings.optimizedImage;
```
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java]: oops there's still a trailing `d` in this word optimizedImage`d`

**Comment by mzorz** [comment]
LGTM - let's just fix that small typo and then feel free to go ahead and merge 👍 @daniloercoli

**Comment by daniloercoli** [comment]
It seems that I'm not able to write "optimizedImage" properly :)

**Comment by mzorz** [comment]
:shipit:
